pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -113,7 +113,8 @@ func identityMatches(set *apps.StatefulSet, pod *v1.Pod) bool {
 		set.Name == parent &&
 		pod.Name == getPodName(set, ordinal) &&
 		pod.Namespace == set.Namespace &&
-		pod.Labels[apps.StatefulSetPodNameLabel] == pod.Name
+		pod.Labels[apps.StatefulSetPodNameLabel] == pod.Name &&
+		pod.Labels[apps.StatefulSetPodOrdinalLabel] == fmt.Sprint(ordinal)
```

Comment:
This will break on upgrade - this check should not be added (it's not backwards compatible when viewing older stateful sets).  We shouldn't be checking this anyway because podName is canonical and ordinal is a secondary attribute of name..

Reply:
Are we sure this will break on upgrade? It looks like the controller updates pod identity if there is not a match:

https://github.com/kubernetes/kubernetes/blob/337e0e18f1aefa199bd0a1786f8eab42e948064c/pkg/controller/statefulset/stateful_pod_control.go#L95-L98

Since I implemented this in the `updateIdentity` function it seemed to make sense to also update `identityMatches`. I'll wait until we can get more feedback on label vs annotation to change this.

Reply:
@liggitt 
> On upgrade, will the controller add the label to existing pods?

I believe so; see above comment.